### DISCUSSION TITLE
How to extract internal links using PyPDF

### DISCUSSION THREAD

**Discussion by swathiJayav** [py-pdf/pypdf#2911] [Q&A] [answered]
I am trying to extract internal links. This is what I have. 

```
    import pypdf

    internal_links = []
    with open(<pdf_path>, 'rb') as pdf_file:
        reader = pypdf.PdfReader(pdf_file)
        for page_num in range(len(reader.pages)):
            page = reader.pages[page_num]
            if "/Annots" in page:
                for annot in page["/Annots"]:
                    annot_obj = annot.get_object()
                    if annot_obj["/Subtype"] == "/Link":
                        if "/Dest" in annot_obj:
                            dest = annot_obj["/Dest"]
                            print(f"page: {page_num} dest: {dest}")
```
I'm unable to get the page number from `dest`. I looked in the pypdf codebase and also in the returned object, but can't find a way to correlate it with a page number.

**Comment by pubpub-zz** [comment]
I convert this thread into a discussion as this is not an issue

**Comment by pubpub-zz** [comment]
this should provide with the page number in your case:
```
pg_num = [ p.indirect_reference for p in reader.flattened_pages].index(dest)
```
be aware that all links are not using directly `["/Dest"]` but some could also use `["/A"]` with `/Goto` actions, and also some documents may store the destinations in the `reader.root_object["/Names"]["/Dests"]`

**Reply by swathiJayav** [reply]
Thank you. 

> but some could also use ["/A"] with /Goto actions,

When does this happen?

> and also some documents may store the destinations in the `reader.root_object["/Names"]["/Dests"]`

When does this happen? `root_object` does not exist on my reader in the current example

**Reply by pubpub-zz** [reply]
> Thank you. 
> 
> > but some could also use ["/A"] with /Goto actions,
> 
> When does this happen?

PDF standard offers alternative solutions. It depends on the program which has been used to create the PDF file

> 
> > and also some documents may store the destinations in the `reader.root_object["/Names"]["/Dests"]`
> 
> When does this happen? `root_object` does not exist on my reader in the current example

Please ensure you are using the latest version

**Reply by swathiJayav** [reply]
We are on 5.0.1 (latest), and still not able to access `reader.root_object["/Names"]["/Dests"]`

**Reply by stefan6419846** [reply]
`reader.root_object` should always exist: https://pypdf.readthedocs.io/en/latest/modules/PdfReader.html#pypdf.PdfReader.root_object If it does not for your PDF file or there are errors, please provide it for further analysis and possibly a stacktrace. Otherwise there is not much to help you with.

**Reply by swathiJayav** [reply]
Actually, I am able to access it, but don't see anything with this key: `reader.root_object["/Names"]`. I'm guessing this only works if `pg_num = [ p.indirect_reference for p in reader.flattened_pages].index(dest)` didn't work. 

Is the recommended way to find the linked page number to try all three of these approaches until one works? Does this cover all the cases we need to check for?

1. `pg_num = [ p.indirect_reference for p in reader.flattened_pages].index(dest)`
2. use ["/A"] with /Goto actions
3. check `reader.root_object["/Names"]["/Dests"] `

**Reply by stefan6419846** [reply, marked as answer]
The following should work:

```python3
dest = annot_obj["/Dest"]
pg_ref = dest[0]
pg_num = [p.indirect_reference for p in reader.flattened_pages].index(pg_ref)
```

For the corresponding syntax, see table 149 of the PDF 2.0 specification.

Full code (slightly modified):

```python3
import pypdf

internal_links = []
with pypdf.PdfReader('Boeing.pdf') as reader:
    for page_num, page in enumerate(reader.pages):
        page = reader.pages[page_num]
        if "/Annots" in page:
            for annot in page["/Annots"]:
                annot_obj = annot.get_object()
                if annot_obj["/Subtype"] == "/Link":
                    if "/Dest" in annot_obj:
                        dest = annot_obj["/Dest"][0]
                        pg_num = [ p.indirect_reference for p in reader.flattened_pages].index(dest)
                        print(f"page: {page_num} pg_num: {pg_num}")
```

And `reader.root_object` works fine here:

```python3
>>> from pypdf import PdfReader
>>> reader = PdfReader('Boeing.pdf')
>>> reader.root_object
{'/Type': '/Catalog', '/Pages': IndirectObject(3, 0, 134954891051504)}
>>> 
```

**Reply by swathiJayav** [reply]
I'm able to get it working when `/Dest` is available, thank you.

But for cases where it's a named link (specified with `/GoTo`), it's not clear how to correlate this with a page number. In the image below, `annot_obj['/A']['/D']` does not match with anything in `reader.root_object["/Names"]["/Dests"] `. Attached is the document as well (taken from pypdf tests). 

![image](https://github.com/user-attachments/assets/e86e3d27-83f5-45f7-b474-58f3aeb36d66)

[pdflatex-outline.pdf](https://github.com/user-attachments/files/17512957/pdflatex-outline.pdf)

**Reply by stefan6419846** [reply]
Retrieve the mapping of names to pages:

```python3
dests = reader.root_object["/Names"]["/Dests"]
kids = dests["/Kids"]
reader.pages[0]  # Populate `reader.flattened_pages`
_pages = [p.indirect_reference for p in reader.flattened_pages]
name_to_page_number = {}
for kid in kids:
    names = kid["/Names"]
    name_count = len(names)
    for i in range(0, name_count, 2):
        page_reference = names[i + 1]["/D"][0]
        name_to_page_number[names[i]] = _pages.index(page_reference)
```

Handle the annotation:

```python3
if "/A" in annot_obj:
    dest = annot_obj["/A"]["/D"]
    pg_num = name_to_page_number[dest]
    print(dest, pg_num)
```

**Reply by swathiJayav** [reply]
This works, thank you so much!

**Comment by swathiJayav** [comment]
Final answer:

For non-named links: https://github.com/py-pdf/pypdf/discussions/2911#discussioncomment-11044334
                     `
For named links: https://github.com/py-pdf/pypdf/discussions/2911#discussioncomment-11049309
